src/access/AccessControl.h

```diff
@@ -31,16 +30,240 @@ namespace Access {
 class AccessControl
 {
 public:
-    /**
-     * Create an access control module. Must be initialized before use, and
-     * deinitialized when finished. Must be configured with an
-     * AccessControlDataProvider, which must outlive this module.
-     */
-    AccessControl(AccessControlDataProvider & dataProvider) : mDataProvider(dataProvider) {}
+    class Entry
+    {
+    public:
+        struct Target
+        {
+            using Flags                        = int;
+            static constexpr Flags kCluster    = 1 << 0;
+            static constexpr Flags kEndpoint   = 1 << 1;
+            static constexpr Flags kDeviceType = 1 << 2;
+            Flags flags;
+            ClusterId cluster;
+            EndpointId endpoint;
+            DeviceTypeId deviceType;
+        };
+
+        class Delegate
```

Comment:
Why do we even need this delegate? This is fairly expensive in terms of vtable cost.

More importantly, you already have a delegate interface defined to retrieve an `Entry` - at that point, the `Entry` itself can be a fairly simple POD right?

Reply:
If we do that we pay the entire cost of Entry wherever we use it. E.g. on the stack. For a type that contains lists. This design avoids that. It's all tradeoffs.

Reply:
The lists themselves wouldn't take up storage within `Entry` right? (i.e it'd be a pointer/Span to some list, which is allocated separately by the provider).

Reply:
That locks the implementation of Entry to use a particular implementation of the ACL. If Entry is polymorphic, then creation is an issue, and new/delete has to be managed, which also means using a smart pointer. That is already what is going on here: think of Entry as a smart pointer (or handle), and Entry::Delegate as the polymorphic implementation.

Another way to look at it: everything is POD storage in this example. You get at it by using a series of ephemeral handle-like objects which know how to affect that POD storage based on the interface APIs you call on them. These ephemeral objects themselves come from pools, and the implementation is smart enough to do things like copy POD from pool to ACL.

At the same time, the interface doesn't require it to be implemented this way, it just allows it. You could write it using STL vector, dynamic allocation, etc.

Unless you mean removing the ability to have different implementations, or at least locking to one implementation at compile time. I suggest that's something we can do as a later transform on the code, if we find it has meaningful gains. Remember this code will only be used from a handful of places: the module itself, its tests, the ZAP code for the cluster that calls it to implement the cluster. (The IM uses this module but just the one check function, it's not managing entries.)